File: py/decode-string.py
'''
  case 1. If we find a letter, add it to the result string.
  case 2. If we find a number, then decode that part of the string until
           find a closing bracket.

Al Gore Rythym For Parsing:
  1. parse the number for the first open bracket
  2. keep track of the bracket count until we find the matching closing bracket
  3. recursively solve that string bycalling decodeString again
  4. The resulting string returned is multiplied by the number of times
    specified and added to the result string
'''
'''
Time Complexity: O(n) since we do the work in a linear pass while constructing
    a string.
Space Complexity: O(n*k) for each set of n characters nested inside a bracket
    with k outside of the bracket.
'''
class Solution0:
    def decodeString(self, s: str) -> str:
        res = ''
        i = 0

        while i < len(s):
            if s[i].isalpha():
                res += s[i]
                i += 1
                continue

            if s[i].isdigit():
                num = 0

                # get all the numbers
                while s[i].isdigit():
                    num = num * 10 + int(s[i])
                    i += 1

                # parse the string inside the bracket
                bracket_count = 1
                i += 1
                inner = ''
                while bracket_count > 0:
                    if s[i] == '[':
                        bracket_count += 1
                    elif s[i] == ']':
                        bracket_count -= 1
                    inner += s[i]
                    i += 1

                # append the parsed expression and continue decoding
                res = res + num * self.decodeString(inner[:-1])
        return res



# With a list to hold the string
class Solution:
    def decodeString(self, s: str) -> str:
        decoded_str_list = []
        i = 0
        while i < len(s):
            if s[i].isdigit():
                j = i + 1

                # Get the length of the the current nested string
                while s[j] != '[':
                    j += 1

                # Get the length of the current string
                k = int(s[i:j])
                i = j + 1
                j = i

                # Track the brackets
                bracket_count = 1
                while True:
                    if s[j] == '[':
                        bracket_count += 1
                    elif s[j] == ']':
                        bracket_count -= 1

                    if bracket_count == 0:
                        break
                    else:
                        j += 1
                decoded_str_list.append(self.decodeString(s[i: j]) * k)
                i = j + 1
            else:
                decoded_str_list.append(s[i])
                i += 1
        return ''.join(decoded_str_list)


# With a stack
class Solution():
    def decodeString(self, s: str) -> str:
        stack = []
        k = 0
        curr_str = ''
        
        for c in s:
            if c == '[':
                stack.append(curr_str)

                # Keep track of k
                stack.append(k)
                curr_str = ''
                k = 0
            elif c == ']':
                # Pop k off the stack
                num = stack.pop()
                prev_str = stack.pop()
                curr_str = prev_str + num * curr_str
            elif c.isdigit():
                k = k*10 + int(c)
            else: # We have a normal character
                curr_str += c
        return curr_str


strings = [
    "2[a2[b]c]",     # abbcabbc
    "3[a]2[bc]",     # aaabcbc
    "3[a2[c]]",      # accaccacc
    "2[abc]3[cd]ef", # abcabccdcdcdef
    ]

for s in strings:
    print(Solution().decodeString(s))

File: cpp/decode-string.cpp
#include <vector>
#include <iostream>
#include <stack>
using namespace std;



class Solution {
public:
    string decodeString(string s) {
        string res;
        const int size = s.size();
        int i = 0;
        
        while (i < size) {
            // Start looping until we find a digit
            while (!isdigit(s[i]) and i < size) {
                res += s[i];
                ++i;
            }

            // If we obtained all the characters without finding a digit, exit
            if (i == size) {
                return res;
            }

            // We found a digit
            int num = 0;
            while (i < size and isdigit(s[i])) {
                num = num * 10 + s[i] - '0';
                ++i;
            }

            // Handle open/closing brackets
            int j = i + 1;
            int bracket_count = 1;
            int k = 1;
            while (j < size and bracket_count > 0) {
                if (s[j] == ']') { --bracket_count; }
                if (s[j] == '[') { ++bracket_count; }
                ++j;
            }

            // Add the temp string and call it recursively on the next subseq.
            // of characters
            string temp = decodeString(s.substr(i + 1, j - i - 2));
            while (num--) {
                res += temp;
            }
            i = j;

        }
        return res;
    }
};



class Solution0 {
public:
    string decodeString(string s) {
        stack<string> string_stack;
        string curr_str;
        
        stack<int> num_stack;
        string num;

        int i = 0;

        for (char c : s) {
            if (c < 58) {
                num.push_back(c);
            }
            else if (c == '[') {
                // append the new string to the top of the stack
                string_stack.push(curr_str);

                // Add the next number
                num_stack.push(stoi(num));

                // clear the old str/num
                curr_str.clear();
                num.clear();
                
            }
            // append the closed sequence the previous-number-seen times
            else if (c == ']') {
                curr_str = string_stack.top() + helper(curr_str,
                                                             num_stack.top());
                string_stack.pop();
                num_stack.pop();
            }
            else {
                curr_str.push_back(c);                
            }
        }
        return curr_str;
    }
    
    // helper function to print characters k amount of times
    string helper(const string &s,const int k) {
        string rs = s;
        for (int i = 1; i < k; ++i) {
            rs += s;
        }
        return rs;
    }
};



int main() {
    vector<string> strings = {
        "2[a2[b]c]",     // abbcabbc
        "3[a]2[bc]",     // aaabcbc
        "3[a2[c]]",      // accaccacc
        "2[abc]3[cd]ef", // abcabccdcdcdef
    };

    for (string s : strings) {
        std::cout << Solution().decodeString(s) << std::endl;
    }
    return 0;
}